atoma-p2p/src/timer.rs

```diff
@@ -0,0 +1,125 @@
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use tokio::{sync::mpsc::UnboundedSender, task::JoinHandle};
+use tracing::{error, instrument};
+
+use crate::service::AtomaP2pNodeError;
+
+/// The interval at which the node will publish usage metrics to the gossipsub topic
+const USAGE_METRICS_HEARTBEAT_INTERVAL: Duration = Duration::from_secs(30);
+
+/// Spawns an asynchronous task that periodically sends usage metrics for nodes
+///
+/// This function creates a background task that:
+/// - For nodes: Periodically collects and sends usage metrics at a fixed interval
+/// - For clients: Immediately returns Ok(()) as clients don't need to publish metrics
+///
+/// The task runs in an infinite loop for nodes, calculating the next heartbeat time
+/// and sending metrics through the provided channel.
+///
+/// # Arguments
+///
+/// * `is_client` - Boolean indicating whether this is a client (true) or node (false)
+/// * `node_public_url` - The public URL of the node
+/// * `node_small_id` - The small ID of the node
+/// * `country` - The country where the node is located
+/// * `usage_metrics_tx` - Unbounded channel sender for transmitting usage metrics
+///
+/// # Returns
+///
+/// Returns a JoinHandle that can be used to monitor or cancel the task
+///
+/// # Errors
+///
+/// Returns AtomaP2pNodeError::UsageMetricsSendError if sending metrics fails
+#[instrument(
+    level = "info",
+    fields(
+        is_client = %is_client,
+        event = "usage_metrics_timer_task",
+        name = "atoma-p2p", 
+    ),
+    skip_all
+)]
+pub fn usage_metrics_timer_task(
+    is_client: bool,
+    node_public_url: String,
+    node_small_id: u64,
+    country: String,
+    usage_metrics_tx: UnboundedSender<NodeUsageMetrics>,
+) -> JoinHandle<Result<(), AtomaP2pNodeError>> {
+    tokio::spawn(async move {
+        // NOTE: We only publish usage metrics for nodes, clients do not need to publish any usage metrics
+        if !is_client {
+            loop {
+                let next_heartbeat =
+                    duration_until_next_heartbeat(USAGE_METRICS_HEARTBEAT_INTERVAL);
```

Comment:
This `duration_until_next_heartbeat` seems a bit excessive since this is a heartbeat system, we can just use a fixed interval sleep

```rust
        tokio::time::sleep(USAGE_METRICS_HEARTBEAT_INTERVAL).await;
```

Reply:
I agree, I will refactor this logic to be simpler and send messages every fixed interval (without any additional node synchronization).